src/app/core/entity-components/entity-list/list-paginator/list-paginator.component.ts

```diff
@@ -0,0 +1,137 @@
+import { EntityMapperService } from "app/core/entity/entity-mapper.service";
+import { getUrlWithoutParams } from "app/utils/utils";
+import { Router } from "@angular/router";
+
+@Component({
+  selector: "app-list-paginator",
+  templateUrl: "./list-paginator.component.html",
+  styleUrls: ["./list-paginator.component.scss"],
+})
+export class ListPaginatorComponent<E extends Entity> {
+  @Input() dataSource: MatTableDataSource<E>;
+
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+
+
+  user: User;
+  paginatorPageSize: number = 10;
+  paginatorPageSizeBeforeToggle: number = 10;
+  paginatorPageIndex: number = 0;
+  showAllToggle: boolean = false;
+
+  // This key is used to save the pagination settings on the user entity
+  readonly paginatorKey: string;
+
+  constructor(
+    private sessionService: SessionService,
+    private entityMapperService: EntityMapperService,
+    private router: Router,
+  ) {
+    this.paginatorKey = getUrlWithoutParams(this.router);
+  }
+
+  ngOnInit() {
+    console.log("PaginatorKey: " + this.paginatorKey);
+    this.user = this.sessionService.getCurrentUser();
+    // Use URl as key to save pagination settings
+    this.paginatorPageSize =
+      this.user.paginatorSettingsPageSize[this.paginatorKey] ||
+      this.paginatorPageSize;
+    this.paginatorPageIndex =
+      this.user.paginatorSettingsPageIndex[this.paginatorKey] ||
+      this.paginatorPageIndex;
+    console.log("OnInit aufgerufen. Length ist " + this.dataSource.data.length);
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    this.showAllToggle =
+      this.paginatorPageSize >= this.dataSource.data.length;
+      console.log("OnChanges aufgerufen. Length ist " + this.dataSource.data.length);
+  }
+
+  ngAfterViewInit() {
+    this.dataSource.paginator = this.paginator;
+    setTimeout(() => {
+      this.paginator.pageIndex = this.paginatorPageIndex;
+      this.paginator.page.next({
+        pageIndex: this.paginator.pageIndex,
+        pageSize: this.paginator.pageSize,
+        length: this.paginator.length,
+      });
+      console.log("AfterViewInit mit Timeout aufgerufen. Length ist " + this.dataSource.data.length);
+    });
+    console.log("AfterViewInit aufgerufen. Length ist " + this.dataSource.data.length);
+  }
+
+  onPaginateChange(event: PageEvent) {
+    this.paginatorPageSize = event.pageSize;
+    this.paginatorPageIndex = event.pageIndex;
+    this.updateUserPaginationSettings();
+    this.showAllToggle =
+      this.paginatorPageSize >= this.dataSource.data.length;
+    console.log("OnPaginateChange aufgerufen. Length ist " + this.dataSource.data.length);
+  }
+
+  getPaginatorPageSizeOptions(): number[] {
+    const ar = [3, 10, 20, 50].filter((n) => {
+      return n < this.dataSource.data.length;
+    });
+    ar.push(this.dataSource.data.length);
+    return ar;
+  }
+
+  getPaginatorPageSize(): number {
+    if (
+      this.dataSource.data.length &&
```

Comment:
what is this check necessary for?

Reply:
This check guarantees that getPaginatorPageSize() does't fail when the data is not yet loaded correctly.